Help Center: Add tailored article list

renancarvalho opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Fixes https://github.com/Automattic/wp-calypso/issues/88707

## Proposed Changes

* This https://github.com/Automattic/wp-calypso/pull/88808 added a feature that allows tailored articles suggestions. This current PR is adding the article's structure to be suggested.
c/c @donalirl 

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Using live link 
* Navigate to the pages (https://github.com/Automattic/wp-calypso/pull/89471/files#diff-1387095f7173d264c294187f77c20df1f285bad3ad7262e615718598da23ed65R1) 
* Open help center
* Make sure the articles suggested are matching the https://github.com/Automattic/wp-calypso/pull/89471/files#diff-1387095f7173d264c294187f77c20df1f285bad3ad7262e615718598da23ed65R1


## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

renancarvalho committed:
Add tailored article list

renancarvalho committed:
Add tailored article list

escapemanuele reviewed (changes requested):
Hey @renancarvalho, I get an error in pages with fewer than 5 posts (like `/import?`

```83029.a0b0c1d….min.js:25 TypeError: tD[a]?.find is not a function```

renancarvalho committed:
Fix structure

renancarvalho committed:
Fix structure

escapemanuele reviewed (changes requested):
Found this!

<img width="429" alt="image" src="https://github.com/Automattic/wp-calypso/assets/52076348/39f5de75-3d24-40dc-a6cb-84f928ce3fd0">

On `/import` we have duplicates. So it seems we're not removing duplicates when we have both tailored articles and search query :) We should do that.

renancarvalho committed:
Filter out duplicated items

renancarvalho committed:
Filter out duplicated items

renancarvalho commented:
> Found this!
> 
> <img alt="image" width="429" src="https://private-user-images.githubusercontent.com/52076348/322005890-39f5de75-3d24-40dc-a6cb-84f928ce3fd0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTI5MzE3NzIsIm5iZiI6MTcxMjkzMTQ3MiwicGF0aCI6Ii81MjA3NjM0OC8zMjIwMDU4OTAtMzlmNWRlNzUtM2QyNC00MGRjLWE2Y2ItODRmOTI4Y2UzZmQwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDEyVDE0MTc1MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTEyMTcxN2Q4NzBiNjdjYWRhOTBjNzMwYWUxYzlhNzcyMDU3YjhmMmQwNDVkMjQ2YWFmNzgxODkwYzM5NDc2M2EmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.V5TjUmS8sXPZgp_D1q_xWmkCfWH00r6clFFWK0TZCWw">
> On `/import` we have duplicates. So it seems we're not removing duplicates when we have both tailored articles and search query :) We should do that.

Great! I have added a `filterOut` function to remove duplicated articles

escapemanuele commented on client/data/help/use-help-search-query.ts:
```diff
@@ -66,8 +74,11 @@ const fetchArticlesAPI = async (
 		} as APIFetchOptions ) ) as SearchResult[];
 	}
+	// Remove articles that are already in the tailored articles.
+	const searchResult = filterOutDuplicatedItems( articlesResponse, searchResultResponse );
```
Pay attention that this is confusing. You pass articles and then search, while the function parameters are search and then articles.
I spent a little trying to understand why it worked, before noticing this swap 😄

escapemanuele commented:
Left a comment regarding parameter naming to be changed :)

One thing that we can think of, in the future... Would it be possible to have only the locale or the blog id? :)

renancarvalho committed:
Fix filter function

renancarvalho committed:
Fix filter function

renancarvalho commented:
> Left a comment regarding parameters naming to be changed :)
> 
> One thing that we can think of, in the future... Would it be possible to have only the locale or the blog id? :)

I think we could think of removing locale since support sites have different blog_id's per “locale”, need more profound investigation, though. Not sure about dev sites.

escapemanuele commented:
Sure thing, let's investigate when we have some down time! 

Opened this issue to track this: https://github.com/Automattic/wp-calypso/issues/89524

escapemanuele reviewed (approved):
Thanks for addressing the feedback! Let's wait for @donalirl and then merge this 👍

donalirl commented:
Overall it's working really well! I checked every screen and the correct pages loaded, except for these ones:

- me/purchases
- me/purchases/billing
- me/purchases/payment-methods
- me/purchases/subscriptions/ (possibly, I don't have this tab)
- marketing/tools

They loaded blank:
![Screenshot 2024-04-16 at 10 36 13](https://github.com/Automattic/wp-calypso/assets/30727666/9749075a-468c-4a80-958e-45e21fe5e446)

Two other things:
* I wasn't able to check the WP-Admin screens, as they don't seem to work within the Calypso testing environment
* Is it possible to set articles for the editor screens? Those are:
  * /page/
  * /post/
  * /wp-admin/site-editor.php

renancarvalho commented:
> * , as they don't seem to work within the Calypso testing environm

Hi @donalirl, thank you for testing this.

I noticed that some IDs don't have a post related to it or to the blog, in this case, the API was failing, I have fixed it, now if the post_Id does not exist it is just ignored. 👍 

Regarding the WP-admin pages, it is possible by updating a testing version of the ETK plugin on your atomic site, I can send you instructions on how to do that.

Finally, yes it should be possible to set articles for /page/, /post/, /wp-admin/site-editor.php, should be the same as the other pages.

🙇

renancarvalho commented:
@donalirl I have tested on my own Atomic site the `wp-admin` URL's

On `wp-admin/edit.php?post_type=page` this is what I get and they match the ids you have added

<img width="420" alt="image" src="https://github.com/Automattic/wp-calypso/assets/2653810/7c84f678-d6ab-437f-8040-ee56f2d2fcb7">


On `/wp-admin/edit.php` also matching the ids you have added

<img width="419" alt="image" src="https://github.com/Automattic/wp-calypso/assets/2653810/103079f5-b091-424b-b53d-4eeb50e88800">


On `wp-admin/post-new.php?post_type=page` also matching the ids you have added

<img width="418" alt="image" src="https://github.com/Automattic/wp-calypso/assets/2653810/52e8c7fb-53c1-43a7-8f9f-445fd009ea63">

On `wp-admin/post-new.php` also matching the ids you have added

<img width="418" alt="image" src="https://github.com/Automattic/wp-calypso/assets/2653810/166772cf-523f-47f8-99fa-dda26d0008f0">

escapemanuele commented:
@donalirl an example of articles not working for me are 276100 and 111341

donalirl commented:
> I noticed that some IDs don't have a post related to it or to the blog, in this case, the API was failing, I have fixed it, now if the post_Id does not exist it is just ignored. 👍

Do you have a record of which IDs those were? It's possible I typed it wrong. For the ones @escapemanuele shared:

276100 - this should be 267100
111341 - this should be 111349

> Finally, yes, it should be possible to set articles for /page/, /post/, /wp-admin/site-editor.php , should be the same as the other pages.

Great, can we add:

```
	"/page/": [ { "locale": "en", "post_ids": [ 147594, 193416, 193554, 167011, 193903 ], "blog_id": 9619154 } ],
	"/post/": [ { "locale": "en", "post_ids": [ 147594, 193416, 193554, 167011, 176675 ], "blog_id": 9619154 } ],
	"/wp-admin/site-editor.php/": [ { "locale": "en", "post_ids": [ 188667, 193554, 162159, 192200, 176675 ], "blog_id": 9619154 } ],

```

renancarvalho committed:
Merge branch 'trunk' into add/helpcenter-tailored-list

renancarvalho committed:
Merge branch 'add/helpcenter-tailored-list' of github.com:Automattic/wp-calypso into add/helpcenter-tailored-list

renancarvalho committed:
Add correct ids

renancarvalho commented:
@donalirl, thank you for providing the new IDs

I have updated the new IDs and added the new pages 👍 
 
The only weird one is `wp-admin/site-editor.php` that took me to this page with the old Help Center icon (FAB if I am not mistaken)

<img width="1001" alt="image" src="https://github.com/Automattic/wp-calypso/assets/2653810/a3cd192d-4a72-49f8-aa3d-2a89b361bc57">

@escapemanuele I think we should take a look at this URL in another task, WDYT? If that is fine, I think we are good to merge this one.

renancarvalho committed:
Add new pages structure

escapemanuele commented:
I think we can deploy this, thanks a lot 👍
